Add rules for automatic differentiation

sethaxen opened the pull request:
This PR adds rules for reverse-mode automatic differentiation (AD). Specifically, it makes the forward and reverse transforms and the integration functions differentiable with Zygote.jl.

This is something I need for my project, and ZygoteRules.jl is a very light dependency (load time increases around 0.2s on my machine with this PR) intended to be used for specifying rules without making Zygote a dependency.

With this PR, you can do things like
```julia
using Hankel, Zygote
q = QDHT(30.0, 100)
fr = exp.(-q.r.^2 ./ 2)
Zygote.gradient(r -> integrateR(exp.(-r.^2 ./ 2), q), q.r)[1]
```
which computes the gradient of the integral wrt the gridpoints.

At some point in the future, the same rules can be migrated to use ChainRulesCore.jl's interface, which will eventually make these rules available to many more AD packages.

sethaxen committed:
Add ZygoteRules as a dependency

sethaxen committed:
Add Zygote rules

sethaxen committed:
Test Zygote rules

chrisbrahms commented:
Sorry for the slow response--had a conference deadline on Friday.

I can't pretend to fully understand how this works since I don't use automatic differentiation myself, but in principle I'm happy to have this. Two things:
1. something has gone wrong with the CI testing. I've tried rerunning the CI but it still won't run. Not sure what's going on there--any ideas?
2. this should be documented, ideally both in the docs (with a quick example) and the readme.

sethaxen committed:
Switch to ChainRules

sethaxen committed:
Add rrule for QDHT constructor

sethaxen committed:
Add frules

sethaxen committed:
Add mutating rrules and frules

sethaxen committed:
Use view

sethaxen committed:
Make symmetric non-mutating

sethaxen committed:
Add reshape method

sethaxen committed:
Make oversample non-mutating

sethaxen committed:
Add inv function

sethaxen committed:
Update dimdot tests to integrate tests

sethaxen committed:
Test mutating and streamline tests

sethaxen committed:
Use oftype

sethaxen committed:
Use ChainRules for test

sethaxen committed:
Fix and test constructor rrule

sethaxen committed:
Add arg for differential of function itself

sethaxen committed:
Add tests for new methods

sethaxen committed:
Document new methods

sethaxen committed:
Reformat

sethaxen committed:
Escape term in docstring

sethaxen commented:
> Sorry for the slow response--had a conference deadline on Friday.

No worries!

> I can't pretend to fully understand how this works since I don't use automatic differentiation myself, but in principle I'm happy to have this.

Great! I'm happy to explain more if you'd like. Autodiff rules are bit nonintuitive at first but then click. The [ChainRules.jl docs](http://www.juliadiff.org/ChainRulesCore.jl/stable/#The-propagators:-pushforward-and-pullback-1) are also nice.

> Two things:
> 
> 1. something has gone wrong with the CI testing. I've tried rerunning the CI but it still won't run. Not sure what's going on there--any ideas?

I don't know. Looks good to me now.

> 2. this should be documented, ideally both in the docs (with a quick example) and the readme.

Absolutely. I'll add an example as well. I switched over to ChainRules since https://github.com/FluxML/Zygote.jl/pull/366, which makes Zygote use ChainRules' rules, is almost complete. So we can hold off merging this PR until that one has been merged.

I also refactored `symmetric`, `Rsymmetric`, and `oversample` to make them autodiffable without any additional rules (basically, use `cat` and `view` instead of mutating). Now also if one wants to oversample many times, the larger `QDHT` can be pre-calculated and used. Now all of the exported functions are compatible with autodiff. Behavior should be the same. I still need to add tests for the forward-mode differentiation rules.

sethaxen committed:
Standardize as array

sethaxen committed:
Safely choose type and size of array

sethaxen committed:
Reformat code

sethaxen committed:
Fix bug in frule

sethaxen committed:
Don't compute unneeded sizes

sethaxen committed:
Test frules

sethaxen committed:
Fix grammar

sethaxen committed:
Add pass-through frule for QDHT constructor

sethaxen committed:
Document the autodiff rules

sethaxen commented:
Okay, this should be feature-complete now, along with documentation. It's only waiting on final review and for https://github.com/FluxML/Zygote.jl/pull/366 to be merged so that the example in the docs works.

sethaxen committed:
Bump ChainRules family compats

sethaxen committed:
Remove mutating rrules

These really aren't safe in any AD system yet

sethaxen committed:
Use prettier NamedTuple syntax

sethaxen committed:
Restrict to latest version of test utils

sethaxen committed:
Drop complex tests

Nothing we do relies on complexness/realness of inputs.

sethaxen committed:
Customize tests for constructor

sethaxen committed:
Remove custom rrule_test

sethaxen committed:
Add comment

sethaxen committed:
Remove conj for guaranteed reals

sethaxen committed:
Update example in docs

sethaxen committed:
Temporarily add Manifest to docs

sethaxen committed:
Remove Zygote compat

sethaxen committed:
Try pointing to repo

sethaxen committed:
Revert "Try pointing to repo"

This reverts commit 520a9064fe742eeed4e3105a6f90f66dfb6447ca.

sethaxen committed:
Revert "Remove Zygote compat"

This reverts commit 5a41bbc0abe21dd6880f96ec94458568a534e87f.

sethaxen committed:
Instantiate first to get Zygote

sethaxen commented:
@chrisbrahms Zygote's master branch now uses ChainRules, so the example in the docs works. This is ready for final review.

chrisbrahms commented on .github/workflows/documenter.yml:
```diff
@@ -16,7 +16,7 @@ jobs:
           version: 1.4
       - name: Install dependencies
-        run: julia --project=docs/ -e 'using Pkg; Pkg.develop(PackageSpec(path=pwd())); Pkg.instantiate()'
+        run: julia --project=docs/ -e 'using Pkg; Pkg.instantiate(); Pkg.develop(PackageSpec(path=pwd()))'
```
What's the reason for this?

sethaxen replied:
I didn't see this documented anywhere, but the original order does not work. It's apparently necessary to do things in this order when using an unregistered package in the docs.

chrisbrahms replied:
Yes that makes sense. In the previous order, `Pkg` would not use the `Manifest.toml` and install the latest release of `Zygote` instead.

chrisbrahms commented on .gitignore:
```diff
@@ -4,5 +4,5 @@
 *.mem
 *.h5
-Manifest.toml
+/Manifest.toml
```
Why do we want to check in the docs `Manifest.toml`?

sethaxen replied:
Currently only Zygote#master uses ChainRules, so we need to check in the manifest until a release is made (Zygote's release schedule is typically a bit slower, so I'd rather not wait). When Zygote does make a release, I'll open a new PR removing the docs manifest.

chrisbrahms reviewed (commented):
Looks good to me. Just a bit confused by the changes to `.gitignore` and the documenter workflow.
